config/config_as_cty.go

```diff
@@ -0,0 +1,195 @@
+package config
+
+import (
+	"encoding/json"
+
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/gocty"
+	ctyjson "github.com/zclconf/go-cty/cty/json"
+
+	"github.com/gruntwork-io/terragrunt/errors"
+	"github.com/gruntwork-io/terragrunt/remote"
+)
+
+// Serialize TerragruntConfig struct to a cty Value that can be used to reference the attributes in other config. Note
+// that we can't straight up convert the struct using cty tags due to differences in the desired representation.
+// Specifically, we want to reference blocks by named attributes, but blocks are rendered to lists in the
+// TerragruntConfig struct, so we need to do some massaging of the data to convert the list of blocks in to a map going
+// from the block name label to the block value.
+func terragruntConfigAsCty(config *TerragruntConfig) (cty.Value, error) {
```

Comment:
Yea I was not very happy with this... but I also couldn't come up with a way to avoid it. I'll see if I can add a test here.

Reply:
Added some unit tests: c5ea857

Reply:
This test does a great job of checking that all the existing fields are converted, but if we were to add a new field, and forget to to add it to this test and to `terragruntConfigAsCty`... The test won't fail, right?

Reply:
Hmm my understanding is that [it should fail here](https://github.com/gruntwork-io/terragrunt/blob/yori-read-terragrunt-config/config/config_as_cty_test.go#L177), where it will exhaust the options when it sees a new struct field.

I'm going to verify this now that I merged in the codegen feature. My expectation is that the test will fail when I rebase and push. If it doesn't, I'll update the test until it fails.

Reply:
Yup failed as expected: https://circleci.com/gh/gruntwork-io/terragrunt/4101

```
=== RUN   TestTerragruntConfigAsCtyDrift
--- FAIL: TestTerragruntConfigAsCtyDrift (0.00s)
    config_as_cty_test.go:178: Unknown struct property: GenerateConfigs
```

Reply:
Ah, nice!